`min_const_generic`  takes precedence over `const_generics`

I expect the code below code to compile, but it fails. `const_generics` should allow more than just the types whitelisted in `min_const_generics`, but it seems rustc only looks at the `min_const_generics` flag in this case.

```Rust
#![feature(const_generics)]
#![feature(min_const_generics)]

struct Contain<const S: &'static [u8]>;
```

The error message is:
```
error: `&'static [u8]` is forbidden as the type of a const generic parameter
 --> src/lib.rs:4:25
  |
4 | struct Contain<const S: &'static [u8]>;
  |                         ^^^^^^^^^^^^^
  |
  = note: the only supported types are integers, `bool` and `char`
  = note: more complex types are supported with `#[feature(const_generics)]`
```

> But why would you want to use the two of them?

Note that:

```rust 
#[feature(const_generics)]
struct S<const I: usize>;
```

results in a [compile error](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=223ce61924e1a2df6825fee6ae0125ee): 

```
error[E0658]: const generics are unstable
 --> src/lib.rs:3:16
  |
3 | struct S<const I: usize>;
  |                ^
  |
  = note: see issue #74878 <https://github.com/rust-lang/rust/issues/74878> for more information
  = help: add `#![feature(min_const_generics)]` to the crate attributes to enable
```

So currently `const_generics` is _not_ a superset of `min_const_generics`.

> > But why would you want to use the two of them?
> 
> Note that:
> 
> ```rust
> #[feature(const_generics)]
> struct S<const I: usize>;
> ```
> 
> results in a [compile error](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=223ce61924e1a2df6825fee6ae0125ee):
> 
> ```
> error[E0658]: const generics are unstable
>  --> src/lib.rs:3:16
>   |
> 3 | struct S<const I: usize>;
>   |                ^
>   |
>   = note: see issue #74878 <https://github.com/rust-lang/rust/issues/74878> for more information
>   = help: add `#![feature(min_const_generics)]` to the crate attributes to enable
> ```
> 
> So currently `const_generics` is _not_ a superset of `min_const_generics`.

You are missing a `!` in your feature. I still think const_generics are a superset of min_const_generics, _but_ rustc just suggest using the smaller one in this case becouse that's the subset aiming to be stabilized for now

--- a/compiler/rustc_ast_passes/src/feature_gate.rs
+++ b/compiler/rustc_ast_passes/src/feature_gate.rs
@@ -608,6 +608,7 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
 
 pub fn check_crate(krate: &ast::Crate, sess: &Session) {
     maybe_stage_features(sess, krate);
+    check_incompatible_features(sess);
     let mut visitor = PostExpansionVisitor { sess, features: &sess.features_untracked() };
 
     let spans = sess.parse_sess.gated_spans.spans.borrow();
@@ -674,6 +675,39 @@ fn maybe_stage_features(sess: &Session, krate: &ast::Crate) {
                 option_env!("CFG_RELEASE_CHANNEL").unwrap_or("(unknown)")
             )
             .emit();
+        }
+    }
+}
+
+fn check_incompatible_features(sess: &Session) {
+    let features = sess.features_untracked();
+
+    let declared_features = features
+        .declared_lang_features
+        .iter()
+        .copied()
+        .map(|(name, span, _)| (name, span))
+        .chain(features.declared_lib_features.iter().copied());
+
+    for (f1, f2) in rustc_feature::INCOMPATIBLE_FEATURES
+        .iter()
+        .filter(|&&(f1, f2)| features.enabled(f1) && features.enabled(f2))
+    {
+        if let Some((f1_name, f1_span)) = declared_features.clone().find(|(name, _)| name == f1) {
+            if let Some((f2_name, f2_span)) = declared_features.clone().find(|(name, _)| name == f2)
+            {
+                let spans = vec![f1_span, f2_span];
+                sess.struct_span_err(
+                    spans.clone(),
+                    &format!(
+                        "features `{}` and `{}` are incompatible, using them at the same time \
+                        is not allowed",
+                        f1_name, f2_name
+                    ),
+                )
+                .help("remove one of these features")
+                .emit();
+            }
         }
     }
 }

--- a/compiler/rustc_feature/src/active.rs
+++ b/compiler/rustc_feature/src/active.rs
@@ -605,3 +605,8 @@ pub const INCOMPLETE_FEATURES: &[Symbol] = &[
     sym::lazy_normalization_consts,
     sym::specialization,
 ];
+
+/// Some features are not allowed to be used together at the same time, if
+/// the two are present, produce an error
+pub const INCOMPATIBLE_FEATURES: &[(Symbol, Symbol)] =
+    &[(sym::const_generics, sym::min_const_generics)];

--- a/compiler/rustc_feature/src/lib.rs
+++ b/compiler/rustc_feature/src/lib.rs
@@ -131,7 +131,7 @@ pub fn find_feature_issue(feature: Symbol, issue: GateIssue) -> Option<NonZeroU3
 }
 
 pub use accepted::ACCEPTED_FEATURES;
-pub use active::{Features, ACTIVE_FEATURES, INCOMPLETE_FEATURES};
+pub use active::{Features, ACTIVE_FEATURES, INCOMPATIBLE_FEATURES, INCOMPLETE_FEATURES};
 pub use builtin_attrs::{
     deprecated_attributes, find_gated_cfg, is_builtin_attr_name, AttributeGate, AttributeTemplate,
     AttributeType, BuiltinAttribute, GatedCfg, BUILTIN_ATTRIBUTES, BUILTIN_ATTRIBUTE_MAP,

--- a/src/test/ui/const-generics/min-and-full-same-time.rs
+++ b/src/test/ui/const-generics/min-and-full-same-time.rs
@@ -1,0 +1,7 @@
+#![feature(const_generics)]
+//~^ ERROR features `const_generics` and `min_const_generics` are incompatible
+#![allow(incomplete_features)]
+#![feature(min_const_generics)]
+
+
+fn main() {}

--- a/src/test/ui/const-generics/min-and-full-same-time.stderr
+++ b/src/test/ui/const-generics/min-and-full-same-time.stderr
@@ -1,0 +1,13 @@
+error: features `const_generics` and `min_const_generics` are incompatible, using them at the same time is not allowed
+  --> $DIR/min-and-full-same-time.rs:1:12
+   |
+LL | #![feature(const_generics)]
+   |            ^^^^^^^^^^^^^^
+...
+LL | #![feature(min_const_generics)]
+   |            ^^^^^^^^^^^^^^^^^^
+   |
+   = help: remove one of these features
+
+error: aborting due to previous error
+

